pkg/client/cache/reflector.go

```diff
@@ -169,14 +249,18 @@ func (r *Reflector) syncWith(items []runtime.Object) error {
 }
 
 // watchHandler watches w and keeps *resourceVersion up to date.
-func (r *Reflector) watchHandler(w watch.Interface, resourceVersion *string, exitWatch <-chan time.Time) error {
+func (r *Reflector) watchHandler(w watch.Interface, resourceVersion *string, exitWatch <-chan interface{}) error {
 	start := time.Now()
 	eventCount := 0
+
+	// Stopping the watcher should be idempotent and if we return from this function there's no way
+	// we're coming back in with the same watch interface.
+	defer w.Stop()
```

Comment:
Agree about the refactor, don't completely agree about the difference between resync and stop. In general I feel like they're different, but the way listAndWatch behaves today, they feel essentially the same. A resync is a cancel unless called from withing a while loop.